Test SPO vote counting

Lucsanszky opened the pull request:
# Description
Follow-up for #4659 
Resolves #4617
Resolves #4727
<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

Lucsanszky committed:
Test SPO voting threshold properties

Resolves #4617

Lucsanszky committed:
Add `delegateSPORewardAddressToDRep_` helper

Lucsanszky committed:
Test SPO default voting during bootstrap

  * Assert that the default vote is `Abstain` in general
  * Assert that `HardForkInitiation` is an exception to the above
    and default vote is `No` in this case

Lucsanszky committed:
Test SPO default voting after bootstrap

  * Assert that default vote is `No` in general
  * Assert that default vote is `No` for `HardForkInitiation`
    and this cannot be overruled by the new default vote customisation
  * Assert that the default vote can be customised by delegating
    a pool's reward account to:
      - An `AlwaysNoConfidence` DRep, in which case the default vote
        becomes `Yes` for a `NoConfidence` action
      - An `AlwaysAbstain` DRep, in which case the default vote
        become `Abstain` in general

aniketd commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs:
```diff
@@ -0,0 +1,351 @@
+                    re {reStakePoolDistr = distr}
+                    gas {gasStakePoolVotes = votes}
+                    (rs ^. rsEnactStateL . ensProtVerL)
+                    `shouldBe` 1
```
It is not easy to read that the two calls to `spoAcceptedRatio` are the same. Perhaps making a common reference will help in making it easier to understand.
```suggestion
              let acceptedRatio =
                    spoAcceptedRatio
                      @era 
                      re {reStakePoolDistr = distr}
                      gas {gasStakePoolVotes = votes}
                      (rs ^. rsEnactStateL . ensProtVerL)
               in if fromCompact totalStake == Coin 0
                     then acceptedRatio `shouldBe` 0
                     else acceptedRatio `shouldBe` 1
                
```

aniketd commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs:
```diff
@@ -0,0 +1,351 @@
+  Gen (TestData era)
+genTestData Ratios {yes, no, abstain, alwaysAbstain, noConfidence} = do
+  let inPools = listOf (arbitrary @(KeyHash 'StakePool (EraCrypto era)))
+  pools <- inPools
```
Why not
```suggestion
  pools <- listOf (arbitrary @(KeyHash 'StakePool (EraCrypto era)))
```

aniketd commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs:
```diff
@@ -0,0 +1,351 @@
+  (a, b, c, d, e, _) <- genPctsOf100
+  pure $ Ratios {yes = a, no = b, abstain = c, alwaysAbstain = d, noConfidence = e}
+
+genPctsOf100 :: Gen (Rational, Rational, Rational, Rational, Rational, Rational)
```
At first I didn't understand why one of the values is being ignored. 😃 👍 
```suggestion
-- | Generates 6 rational values for voting ratios. One of them should be ignored to account for stake that didn't participate.
genPctsOf100 :: Gen (Rational, Rational, Rational, Rational, Rational, Rational)
```

aniketd reviewed (commented):
Just a few minor suggestions. LGTM! 👍

lehins commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs:
```diff
@@ -0,0 +1,351 @@
+      [([KeyHash 'StakePool (EraCrypto era)], a)] ->
+      Map (KeyHash 'StakePool (EraCrypto era)) a
+    unionAllFromLists =
+      foldr (Map.union . (\(l, v) -> Map.fromList [(cred, v) | cred <- l])) Map.empty
```
This simplification uses [`fromKeys` function that we have available](https://github.com/IntersectMBO/cardano-ledger/blob/97c2c87a710d6b122cee6a37f47c0aa952ae7bef/libs/cardano-data/src/Data/MapExtras.hs#L251C1-L251C9) as well as the Monoid instance of `Map`
```suggestion
    unionAllFromLists = foldMap (\(ks, v) -> fromKeys (const v) ks)
```

lehins commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs:
```diff
@@ -0,0 +1,351 @@
+    mkDelegatees drep =
+      Map.fromList
+        . map (\(_, params) -> (raCredential $ ppRewardAccount params, drep))
+        . Map.toList
```
This simplification uses [`fromKeys` function that we have available](https://github.com/IntersectMBO/cardano-ledger/blob/97c2c87a710d6b122cee6a37f47c0aa952ae7bef/libs/cardano-data/src/Data/MapExtras.hs#L251C1-L251C9)
```suggestion
      fromKeys (const drep) . map (raCredential . ppRewardAccount) . Map.elems
```

lehins commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs:
```diff
@@ -0,0 +1,351 @@
+
+    genPoolParams p = do
+      params <- arbitrary
+      pure $ Map.fromList [(cred, params) | cred <- p]
```
This simplification uses [`fromKeys` function that we have available](https://github.com/IntersectMBO/cardano-ledger/blob/97c2c87a710d6b122cee6a37f47c0aa952ae7bef/libs/cardano-data/src/Data/MapExtras.hs#L251C1-L251C9)

```suggestion
      pure $ fromKeys (const params) p
```

However, both my suggestion above and the implementation will use the same pool params, which seems wrong to me. Not only because it didn't look like that is what we need in the tests, but also because with this generation the invariant is violated where params contain the same poolId as the key in the Map.

It seems to me this was the intended generator:
```suggestion
      let genPoolParams poolId = do
            poolParams <- arbitrary
            pure $ poolParams { ppId = poolId }
      sequence $ fromKeys genPoolParams p
```

I can't fully understand the tests yet to tell for sure which one of these is the desired generators.

lehins commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs:
```diff
@@ -0,0 +1,351 @@
+      , stakeNoConfidence = Coin . fromIntegral $ length poolsNoConfidence
+      , stakeNotVoted = Coin . fromIntegral $ length rest
+      , delegatees = Map.union delegateesAA delegateesNC
+      , poolParams = Map.union poolParamsRest $ Map.union poolParamsAA poolParamsNC
```
```suggestion
      , poolParams = Map.unions [poolParamsRest, poolParamsAA, poolParamsNC]
```

Lucsanszky commented:
Note to whoever takes this over as I'm off this week: I reinstated the previously removed tests in the last commit. I only adjusted them minimally, so they pass. I think the one with the `NoConfidence` action could be removed as I often used `NoConfidence` in the Imp tests to check the vote counting behaviour. The committee update one could stay or maybe adjusted to better fit the recently added tests. I also wouldn't be against dropping that commit altogether and not reinstating those old tests at all because committee updates are also covered by the newly added `acceptedRatioProp` as that could generate such cases.

lehins commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs:
```diff
@@ -0,0 +1,345 @@
+          rs
+          gas
+          `shouldBe` votingStakePoolThreshold @era rs (gasAction gas)
+          == SJust minBound
```
Minor suggestion to improve readability:
```suggestion
        let re' = re {reStakePoolDistr = PoolDistr Map.empty (fromJust . toCompact $ Coin 100)}
             in spoAccepted @era re' rs gas
                  `shouldBe` 
                  (votingStakePoolThreshold @era rs (gasAction gas) == SJust minBound)
```

Lucsanszky commented:
Regarding #4727, I dropped the `NoConfidence` test since we cover that case quite extensively and I changed one of the newly added tests to cover the `CommitteeUpdate` case. I think this is now good to go, so I'll enable auto-merge and wait for approval.

lehins reviewed (approved):
:+1: 
Thank you!
